### PR TITLE
CDAP-4133 Ability to get the hostname of a appfabric service as a part of live-info.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://issues.cask.co/browse/CDAP-4133

live-info for the app fabric now returns the following information instead of empty object:
```java
{
    "containers": [
        {
            "container": "N/A", 
            "host": "x.y.z", 
            "instance": 0, 
            "memory": 0, 
            "name": "appfabric", 
            "type": "system_service", 
            "virtualCores": 0
        }
    ]
}
```

This allows us to locate the leader in HA setup.

### CODE REVIEW

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-RB304-1

**Comment by chtyim** [comment]
Why memory is 0 and instance is 0? Also, for container, if it is not available, shouldn't it be left out / having null instead of string "N/A"?

**Comment by sagarkapare** [comment]
Since app fabric does not run in the container I added "N/A" for the container and 0 for memory, instance and virtualCores.

Should this simply return:
 
```java
{
    "containers": [
        {
            "host": "x.y.z", 
            "name": "appfabric", 
            "type": "system_service"
        }
    ]
}
```

**Comment by chtyim** [comment]
I think that would be better, as long as UI can handle it correctly.

**Comment by sagarkapare** [comment]
Sounds good. I will check from UI side as well.

**Comment by sagarkapare** [comment]
@chtyim Updated the code to not return the container related information for the app fabric service. Please take a look again. Thanks!

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/AppFabricServiceRuntimeModule.java [hunk 1]
```diff
@@ -143,7 +144,7 @@ protected void configure() {
                                mapBinder.addBinding(Constants.Service.METRICS)
                                         .to(InMemoryMetricsServiceManager.class);
                                mapBinder.addBinding(Constants.Service.APP_FABRIC_HTTP)
-                                        .to(AppFabricServiceManager.class);
+                                        .to(InMemoryAppFabricServiceManager.class);
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/AppFabricServiceRuntimeModule.java]: may be we can reuse this map, as standalone and in-memory both use the same map entries.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  @Override
+  public boolean canCheckStatus() {
+    return true;
+  }
+
+  @Override
+  public boolean isServiceAvailable() {
+    return true;
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: so.. this is always true, as we can reach this endpoint only when app.fabric is up, as MonitorHandler runs as a part of app.fabric ?

File: cdap-proto/src/main/java/co/cask/cdap/proto/Containers.java [hunk 3]
```diff
@@ -46,15 +46,15 @@
...
     private final String host;
-    private final int memory;
-    private final int virtualCores;
+    private final Integer memory;
+    private final Integer virtualCores;
     private final Integer debugPort;
 
-    public ContainerInfo(ContainerType type, String name, int instance, String container,
-                  String host, int memory, int virtualCores, Integer debugPort) {
+    public ContainerInfo(ContainerType type, String name, Integer instance, String container,
```
Comment (gsps1) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Containers.java]: should add ``@Nullable``

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+/**
+ * App Fabric Service Management in Distributed Mode.
+ */
+public class AppFabricServiceManager extends AbstractDistributedMasterServiceManager {
+
+  private final InetAddress hostname;
+
+  @Inject
+  public AppFabricServiceManager(CConfiguration cConf, TwillRunnerService twillRunnerService,
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: you can implement ``MasterServiceManager`` and skip injecting unnecessary objects like ``TwillRunnerService``, etc.

**Comment by gsps1** [comment]
few minor comments, otherwise code changes LGTM, also make sure you are able to deserialize as expected with your new changes and also add documentation section about discovering master service in HA environment.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/AppFabricServiceRuntimeModule.java (hunk 1 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/AppFabricServiceRuntimeModule.java]: Moved creation of the map to the common method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: correct.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Containers.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Containers.java]: Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java (hunk 4 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: Fixed.

**Comment by sagarkapare** [comment]
@gsps1 Addressed review comment and also verified that the deserialization of the returned JSON happens correctly. Will add doc changes as another PR. Please take a look again. Thanks!

**Comment by gsps1** [comment]
LGTM

**Pull request merged**